engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/HashCodeAnyNode.java

```diff
@@ -0,0 +1,258 @@
+import com.oracle.truffle.api.profiles.ValueProfile;
+import java.time.LocalDateTime;
+import java.time.ZonedDateTime;
+import java.util.Arrays;
+import org.enso.interpreter.dsl.AcceptsError;
+import org.enso.interpreter.dsl.BuiltinMethod;
+import org.enso.interpreter.runtime.callable.UnresolvedConversion;
+import org.enso.interpreter.runtime.callable.UnresolvedSymbol;
+import org.enso.interpreter.runtime.error.WarningsLibrary;
+
+/**
+ * Hashing contract:
+ * Whenever two objects are equal ({@code EqualsAnyNode} returns {@code true}), their hashcode
+ * should equal.
+ * More formally:
+ * {@code For all objects o1, o2: if o1 == o2 then hash(o1) == hash(o2)}
+ */
+@BuiltinMethod(
+    type = "Any",
+    name = "hash_code",
+    description = "Returns has code of the object."
+)
+@GenerateUncached
+public abstract class HashCodeAnyNode extends Node {
+
+  public static HashCodeAnyNode build() {
+    return HashCodeAnyNodeGen.create();
+  }
+
+  public abstract long execute(@AcceptsError Object self);
+
+  /** Specializations for primitive values **/
+
+  @Specialization
+  long hashCodeForLong(long l) {
+    return l;
+  }
+
+  @Specialization
+  long hashCodeForInt(int i) {
+    return i;
+  }
+
+  @Specialization
+  long hashCodeForDouble(double d) {
+    // TODO: More precise conversion?
+    return Double.valueOf(d).longValue();
+  }
+
+  @Specialization
+  long hashCodeForUnresolvedSymbol(UnresolvedSymbol selfSymbol) {
+    throw new UnsupportedOperationException("unimplemented");
+  }
+
+  @Specialization
+  long hashCodeForUnresolvedConversion(UnresolvedConversion selfConversion) {
+    throw new UnsupportedOperationException("unimplemented");
+  }
+
+  @Specialization(guards = {
+      "warnLib.hasWarnings(selfWithWarning)"
+  }, limit = "3")
+  long hashCodeForWarning(Object selfWithWarning,
+      @CachedLibrary("selfWithWarning") WarningsLibrary warnLib,
+      @Cached HashCodeAnyNode hashCodeNode
+  ) {
+    try {
+      return hashCodeNode.execute(
+          warnLib.removeWarnings(selfWithWarning)
+      );
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /** Specializations for interop values **/
+
+  @Specialization(guards = {
+      "interop.isBoolean(selfBool)"
+  }, limit = "3")
+  long hashCodeForBooleanInterop(Object selfBool,
+      @CachedLibrary("selfBool") InteropLibrary interop) {
+    try {
+      return Boolean.hashCode(
+          interop.asBoolean(selfBool)
+      );
+    } catch (UnsupportedMessageException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  @Specialization(guards = {
+      "!interop.isDate(selfTimeZone)",
+      "!interop.isTime(selfTimeZone)",
+      "interop.isTimeZone(selfTimeZone)",
+  }, limit = "3")
+  long hashCodeForTimeZoneInterop(Object selfTimeZone,
+      @CachedLibrary("selfTimeZone") InteropLibrary interop) {
+    try {
+      return interop.asTimeZone(selfTimeZone).hashCode();
```

Comment:
Will this PE well? Isn't the `TimeZone.hashCode()` may have multiple implementations, I am afraid it will have to be behind `@TruffleBoundary`.

Reply:
No, added `@TruffleBoundary`.